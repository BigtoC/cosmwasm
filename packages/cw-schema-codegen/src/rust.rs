use askama::Template;

pub struct EnumVariantTemplate<'a> {
    pub name: &'a str,
    pub ty: TypeTemplate<'a>,
}

#[derive(Template)]
#[template(escape = "none", path = "rust/enum.tpl.rs")]
pub struct EnumTemplate<'a> {
    pub name: &'a str,
    pub variants: &'a [EnumVariantTemplate<'a>],
}

pub struct FieldTemplate<'a> {
    pub name: &'a str,
    pub ty: &'a str,
}

pub enum TypeTemplate<'a> {
    Unit,
    Tuple(&'a [&'a str]),
    Named {
        fields: &'a [FieldTemplate<'a>],
    }
}

#[derive(Template)]
#[template(escape = "none", path = "rust/struct.tpl.rs")]
pub struct StructTemplate<'a> {
    pub name: &'a str,
    pub ty: TypeTemplate<'a>,
}
