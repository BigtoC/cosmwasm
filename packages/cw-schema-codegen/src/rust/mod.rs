use self::template::{
    EnumTemplate, EnumVariantTemplate, FieldTemplate, StructTemplate, TypeTemplate,
};
use std::{borrow::Cow, io};

pub mod template;

fn expand_node_name<'a>(
    schema: &'a cw_schema::SchemaV1,
    node: &'a cw_schema::Node,
) -> Cow<'a, str> {
    match node.value {
        cw_schema::NodeType::Array { items } => {
            let items = &schema.definitions[items];
            format!("Vec<{}>", expand_node_name(schema, items)).into()
        }
        cw_schema::NodeType::Float => "f32".into(),
        cw_schema::NodeType::Double => "f64".into(),
        cw_schema::NodeType::Boolean => "bool".into(),
        cw_schema::NodeType::String => "String".into(),
        cw_schema::NodeType::Integer { signed, precision } => {
            let ty = if signed { "i" } else { "u" };
            format!("{ty}{precision}").into()
        }
        cw_schema::NodeType::Binary => "Vec<u8>".into(),
        cw_schema::NodeType::Optional { inner } => {
            let inner = &schema.definitions[inner];
            format!("Option<{}>", expand_node_name(schema, inner)).into()
        }
        cw_schema::NodeType::Struct(..) => node.name.as_ref().into(),
        cw_schema::NodeType::Tuple { ref items } => {
            let items = items
                .iter()
                .map(|item| expand_node_name(schema, &schema.definitions[*item]))
                .collect::<Vec<_>>()
                .join(", ");

            format!("({})", items).into()
        }
        cw_schema::NodeType::Enum { .. } => node.name.as_ref().into(),
        _ => todo!(),
    }
}

fn prepare_docs(desc: Option<&str>) -> Cow<'_, [Cow<'_, str>]> {
    desc.map(|desc| {
        desc.lines()
            .map(|line| line.replace('"', "\\\"").into())
            .collect()
    })
    .unwrap_or(Cow::Borrowed(&[]))
}

pub fn process_node<O>(
    output: &mut O,
    schema: &cw_schema::SchemaV1,
    node: &cw_schema::Node,
) -> io::Result<()>
where
    O: io::Write,
{
    match node.value {
        cw_schema::NodeType::Struct(ref sty) => {
            let structt = StructTemplate {
                name: &node.name,
                docs: prepare_docs(node.description.as_deref()),
                ty: match sty {
                    cw_schema::StructType::Unit => TypeTemplate::Unit,
                    cw_schema::StructType::Named { ref properties } => TypeTemplate::Named {
                        fields: properties
                            .iter()
                            .map(|(name, prop)| {
                                let ty = expand_node_name(schema, &schema.definitions[prop.value]);
                                FieldTemplate {
                                    name: Cow::Borrowed(name),
                                    docs: prepare_docs(prop.description.as_deref()),
                                    ty,
                                }
                            })
                            .collect(),
                    },
                    _ => unreachable!(),
                },
            };

            writeln!(output, "{structt}")?;
        }
        cw_schema::NodeType::Enum { ref cases, .. } => {
            let enumm = EnumTemplate {
                name: &node.name,
                docs: prepare_docs(node.description.as_deref()),
                variants: cases
                    .iter()
                    .map(|(name, case)| EnumVariantTemplate {
                        name,
                        docs: prepare_docs(case.description.as_deref()),
                        ty: match case.value {
                            cw_schema::EnumValue::Unit => TypeTemplate::Unit,
                            cw_schema::EnumValue::Tuple { ref items } => {
                                let items = items
                                    .iter()
                                    .map(|item| {
                                        let node = &schema.definitions[*item];
                                        expand_node_name(schema, node)
                                    })
                                    .collect();

                                TypeTemplate::Tuple(items)
                            }
                            cw_schema::EnumValue::Named { ref properties, .. } => {
                                TypeTemplate::Named {
                                    fields: properties
                                        .iter()
                                        .map(|(name, prop)| {
                                            let ty = expand_node_name(
                                                schema,
                                                &schema.definitions[prop.value],
                                            );
                                            FieldTemplate {
                                                name: Cow::Borrowed(name),
                                                docs: prepare_docs(prop.description.as_deref()),
                                                ty,
                                            }
                                        })
                                        .collect(),
                                }
                            }
                            _ => unreachable!(),
                        },
                    })
                    .collect(),
            };

            writeln!(output, "{enumm}")?;
        }
        _ => (),
    }

    Ok(())
}
