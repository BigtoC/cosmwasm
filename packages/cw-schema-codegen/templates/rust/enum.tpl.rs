pub enum {{ name }} {
    {% for variant in variants %}
        {{ variant.name }}
        {% match variant.ty %}
            {% when TypeTemplate::Unit %}
            {% when TypeTemplate::Tuple with (types) %}
                (
                    {% for ty in types %}
                        {{ ty }},
                    {% endfor %}
                )
            {% when TypeTemplate::Named with { fields } %}
                {
                    {% for field in fields %}
                        {{ field.name }}: {{ field.ty }},
                    {% endfor %}
                }
        {% endmatch %}
        ,
    {% endfor %}
}