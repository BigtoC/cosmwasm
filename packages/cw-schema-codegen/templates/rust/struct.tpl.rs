{% for doc in docs %}
    #[doc = "{{ doc }}"]
{% endfor %}

pub struct {{ name }}

{% match ty %}
    {% when TypeTemplate::Unit %}
        ;
    {% when TypeTemplate::Tuple with (types) %}
        (
            {% for ty in types %}
                {{ ty }},
            {% endfor %}
        );
    {% when TypeTemplate::Named with { fields } %}
        {
            {% for field in fields %}
                {% for doc in field.docs %}
                    #[doc = "{{ doc }}"]
                {% endfor %}

                {{ field.name }}: {{ field.ty }},
            {% endfor %}
        }
{% endmatch %}
