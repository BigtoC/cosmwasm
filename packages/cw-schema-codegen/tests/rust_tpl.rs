use askama::Template;
use cw_schema_codegen::rust::{EnumTemplate, EnumVariantTemplate};

#[test]
fn simple_enum() {
    let tpl = EnumTemplate {
        name: "Simple",
        variants: &[
            EnumVariantTemplate {
                name: "One",
                types: None,
            },
            EnumVariantTemplate {
                name: "Two",
                types: None,
            },
        ],
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}

#[test]
fn complex_enum() {
    let tpl = EnumTemplate {
        name: "Complex",
        variants: &[
            EnumVariantTemplate {
                name: "One",
                types: Some(&["u64"]),
            },
            EnumVariantTemplate {
                name: "Two",
                types: Some(&["String", "u64"]),
            },
        ],
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}

#[test]
fn empty_enum() {
    let tpl = EnumTemplate {
        name: "Empty",
        variants: &[],
    };

    let rendered = tpl.render().unwrap();
    insta::assert_snapshot!(rendered);
}
