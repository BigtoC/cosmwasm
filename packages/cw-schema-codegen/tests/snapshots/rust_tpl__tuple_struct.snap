---
source: packages/cw-schema-codegen/tests/rust_tpl.rs
expression: rendered
---
    #[doc = "Tuple struct"]


pub struct Tuple


        (
            
                u64,
            
                String,
            
        );
